### DISCUSSION TITLE
Neural Audio Codec with Continuous Vectors

### DISCUSSION THREAD

**Discussion by manmay-nakhashi** [lucidrains/naturalspeech2-pytorch#8] [Ideas]
@lucidrains  This is something different from Enocodec and SoundStream 
We use a neural audio codec to convert speech waveform into continuous vectors instead of discrete tokens
and 
The audio encoder consists of several convolutional blocks with a
total downsampling rate of 200 for 16KHz audio.
means they are compressing 16000 khz audio to 80 size continuous vector.

**Comment by lucidrains** [comment]
@manmay-nakhashi yea, they do assert that, but never showed any experiments comparing the two

what they actually did in the paper follows all the other recent successes. they used the soundstream architecture with the residual VQ, and even had a special loss to each quantizer codebook

<img width="822" alt="Screen Shot 2023-04-29 at 10 49 18 AM" src="https://user-images.githubusercontent.com/108653/235316897-559435c5-f958-4cf9-8e51-e9c66f3ec514.png">

<img width="846" alt="Screen Shot 2023-04-29 at 10 49 34 AM" src="https://user-images.githubusercontent.com/108653/235316905-bb4593f5-3073-4473-a2dc-44a80d916cc5.png">

**Reply by manmay-nakhashi** [reply]
@lucidrains  so computation of  continuous vectors is the key and that depends on 
![cont_vec_eq](https://user-images.githubusercontent.com/17176700/235320642-bef626ef-94b4-45a9-8433-59cfc7fad2f9.png)
which can be computed from  codebook embeddings and the quantized token IDs

will it be something like this ?

# Residual Vector Quantizer
```
z = []
for i in range(0, h.size(2), hop_size):
    frame = h[:, :, i:i+hop_size]

    # Calculate L2 distance between frame and codebooks
    distances = torch.norm(frame.unsqueeze(1) - self.codebooks, dim=-1, p=2)

    # Find closest codebook index
    indices = torch.argmin(distances, dim=1)

    # Gather codebook embeddings based on indices
    embeddings = self.codebooks[torch.arange(self.num_res_vq_blocks), indices]

    # Calculate residual vectors
    residuals = frame - embeddings

    # Concatenate residual vectors
    z.append(residuals.reshape(residuals.size(0), -1))

z = torch.stack(z, dim=1)
```
where we concatinate vq over timre

**Reply by lucidrains** [reply]
yup, all that is already taken care of here https://github.com/lucidrains/vector-quantize-pytorch#residual-vq
